File: python/ops_gnn/ind2ptr.py
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

"""Public ind2ptr API."""

from . import _pybind
from .typing import Tensor


def ind2ptr(ind: Tensor, num_rows: int) -> Tensor:
    """Convert sorted row indices to a CSR row-pointer tensor.

    Drop-in replacement for ``torch.ops.torch_sparse.ind2ptr(ind, M)`` on Ascend NPU.
    The second argument is the row count (historically named ``M`` in torch_sparse).

    Launches asynchronously on the current NPU stream (same as pytorch_sparse CUDA).
    Synchronize with ``torch.npu.synchronize()`` / ``torch.npu.current_stream().synchronize()``
    before reading results on host if needed.

    Args:
        ind (Tensor): 1-D ``torch.long`` tensor of non-decreasing row indices on NPU.
        num_rows (int): Number of rows. Output length is ``num_rows + 1``.

    Returns:
        Tensor: 1-D ``torch.long`` CSR pointer of shape ``[num_rows + 1]`` on NPU.

    Example:
        >>> import torch
        >>> import ops_gnn
        >>> row = torch.tensor([2, 2, 4, 5, 5, 6], dtype=torch.long, device='npu')
        >>> ops_gnn.ind2ptr(row, 8)
        tensor([0, 0, 0, 2, 2, 3, 5, 6, 6], device='npu:0')
    """
    if not ind.is_contiguous():
        ind = ind.contiguous()
    return _pybind.ind2ptr(ind, int(num_rows))

File: csrc/npu/ind2ptr/op_host/ind2ptr.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "ind2ptr/op_host/ind2ptr.h"
#include "ind2ptr/op_kernel/arch35/ind2ptr.h"

#include "torch_npu/csrc/core/npu/NPUGuard.h"
#include "torch_npu/csrc/core/npu/NPUStream.h"

// Launch AscendC kernel on the current NPU stream and return (async).
torch::Tensor ind2ptr(torch::Tensor ind, int64_t M)
{
    TORCH_CHECK(ind.dim() == 1, "ind2ptr: ind must be 1-D");
    TORCH_CHECK(ind.scalar_type() == at::kLong, "ind2ptr: ind must be int64 (torch.long)");
    TORCH_CHECK(M >= 0, "ind2ptr: M must be non-negative");

    c10_npu::OptionalNPUGuard device_guard(ind.device());

    auto out = torch::empty({M + 1}, ind.options());
    if (ind.numel() == 0) {
        return out.zero_();
    }

    if (!ind.is_contiguous()) {
        ind = ind.contiguous();
    }

    auto stream = c10_npu::getCurrentNPUStream(ind.device().index());
    Ind2Ptr(ind.data_ptr<int64_t>(), out.data_ptr<int64_t>(), M, ind.numel(), stream);
    return out;
}
